clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -156,13 +161,27 @@ public void ensureValid() {
     }
 
     /**
-     * Get the timestamp of the first record in this batch. It is always the create time of the record even if the
+     * Gets the base timestamp of the batch which is used to calculate the timestamp deltas.
+     * 
+     * @return The base timestamp or
+     *         {@link RecordBatch#NO_TIMESTAMP} if the batch is empty
+     */
+    public long baseTimestamp() {
```

Comment:
I think we need to update `DefaultRecordBatch.RecordIterator` to use `baseTimestamp()` instead of `firstTimestamp()`. We should also make sure we have a test case to ensure that the record timestamps remain value even when the delete horizon has been set.

Reply:
I've updated RecordIterator. I also expanded on `MemoryRecordsTest.testBaseTimestampToDeleteHorizonConversion` to check the record timestamp to verify it is the correct value.